### ISSUE TITLE
Pyright errors on subscripting `Union[ dict, < other types >]`

### ISSUE THREAD

**Issue by gegnew** [microsoft/pyright#2218]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

**Describe the bug**
When describing a custom JSON type, pyright can't handle subscripting of a Union type.

**To Reproduce**
Following [this method](https://github.com/python/typing/issues/182#issuecomment-899624078) for defining a JSON type, see code example below.

**Expected behavior**
`mypy` has no issue with this definition, and I'd expect `pyright` to be okay with it as well.

**Screenshots or Code**
If applicable, add screenshots or the text of the code (surrounded by triple back ticks) to help explain your problem.
```python
from __future__ import annotations

from typing import Union, Protocol, Type, TYPE_CHECKING

if TYPE_CHECKING:
    class JSONArray(list[JSON], Protocol):  # type: ignore
        __class__: Type[list[JSON]]  # type: ignore

    class JSONObject(dict[str, JSON], Protocol):  # type: ignore
        __class__: Type[dict[str, JSON]]  # type: ignore

    JSON = Union[None, float, str, JSONArray, JSONObject]



d: JSONObject = {"a": "b"} # [Pyright reportGeneralTypeIssues]
# [E] Expression of type "dict[str, str]" cannot be assigned to declared type "JSONObject" 
# "__class__" is an incompatible type   
# Type "property" cannot be assigned to type "Type[dict[str, JSON]]"
```

**VS Code extension or command-line**
`pyright 1.1.162` via `coc-pyright`

**Additional context**

**Comment by vanburgerberg** [comment]
A simpler example:
```python
from typing import Protocol, Type


class Proto(Protocol):
    __class__: Type[dict[str, str]]


foo = {"a": "b"}

# the types are the same
reveal_type(foo.__class__)
reveal_type(Proto.__class__)

d: Proto = foo
# error: Expression of type "dict[str, str]" cannot be
# assigned to declared type "Proto"
# "dict[str, str]" is incompatible with protocol "Proto"
# "__class__" is an incompatible type
# Type "property" cannot be assigned to type
# "Type[dict[str, str]]" (reportGeneralTypeIssues)
```
It seems that Pyright doesn't behave as expected when `__class__` is defined in the protocol body

**Comment by erictraut** [comment]
Defining a `__class__` in a protocol is pretty non-standard. I've never seen this pattern before. 

There are a couple of reasons this doesn't work in pyright currently.

The `__class__` attribute is defined in builtins.pyi as a property, not a simple instance variable. Mypy sometimes treats these the same when doing protocol comparisons. Pyright treats them as distinct because a property has somewhat different semantics than an instance variable, so it's not generally safe to replace one with the other.

If we change the code to replace the instance variable in the protocol with properties, it looks like this:

```python
from __future__ import annotations

from typing import Union, Protocol, Type, TYPE_CHECKING

if TYPE_CHECKING:

    class JSONArray(list[JSON], Protocol):  # type: ignore
        @property
        def __class__(self) -> Type[list[JSON]]:
            ...

        @__class__.setter
        def __class__(self, x: Type[list[JSON]]):
            ...

    class JSONObject(dict[str, JSON], Protocol):  # type: ignore
        @property
        def __class__(self) -> Type[dict[str, JSON]]:
            ...

        @__class__.setter
        def __class__(self, x: Type[dict[str, JSON]]):
            ...

    JSON = Union[None, float, str, JSONArray, JSONObject]

d: JSONObject = {"a": "b"}
```

This still doesn't type check, however, because pyright doesn't apply bidirectional type inference to the `__class__` type declaration. So the normal type inference rules apply, and the type of the expression `{"a": "b"}` is inferred to be `dict[str, str]`, which isn't compatible with `dict[str, JSON]`. To work around that, you would need to replace the last line with the following:

```python
x: dict[str, JSON] = {"a": "b"}
d: JSONObject = x
```

I'll also point out that you're creating a class that subclasses from protocol and another class, which isn't supported in the runtime, hence your need to add `# type: ignore` comments.

So pyright is working as I would expect here. 

Why are you using `__class__` in your protocol in this case? Have you seen this done before? Like I said above, I've never seen this pattern used before.

I'm guessing that you are doing these unusual contortions to work around mypy's lack of support for recursive type aliases. In pyright, here's the recommended approach:

```python
from typing import Union

JSON = Union[None, float, str, "JSONArray", "JSONObject"]
JSONArray = list[JSON]
JSONObject = dict[str, JSON]

d: JSONObject = {"a": "b"}
```

**Comment by erictraut** [comment]
I'm going to close this issue because I think pyright is working as expected in this case.
